### DISCUSSION TITLE
react-codemod + yarn berry issues

### DISCUSSION THREAD

**Discussion by wdfinch** [yarnpkg/berry#2997] [Q&A] [answered]
Hi, I'm trying to run some scripts from the react-codemod library. I'm currently getting this error on `3.0.0-rc.5`.

```
Error: react-codemod tried to access .bin, but it isn't declared in its dependencies; this makes the require call ambiguous and unsound.

Required package: .bin (via ".bin/jscodeshift")
Required by: react-codemod@npm:5.4.3 (via /home/wfinch/work/dialsource1/dialsource/frontend/.yarn/cache/react-codemod-npm-5.4.3-df3423e618-4e26e191c3.zip/node_modules/react-codemod/bin/)
```

I understand the problem, however I'm not sure exactly how to fix this one because it seems the library is trying to access `.bin/jscodeshift` rather than just `jscodeshift`

I did add the following to my .yarnrc file, however, it did not fix the issue.

```
packageExtensions:
  "react-codemod@*":
    dependencies:
      "jscodeshift": "*"
```

How would I solve this? Would I need to unplug the package or use some special naming conversion? I tried using `.bin/jscodeshift` as well in .yarnrc and no luck.

**Comment by merceyz** [comment, marked as answer]
It's an issue with how they try to locate the binary for `jscodeshift` (https://github.com/reactjs/react-codemod/blob/b34b92a1f0b8ad333efe5effb50d17d46d66588b/bin/cli.js#L20), Yarn in PnP mode will never produce that file so even if we allowed undeclared requires for dotfiles (`.bin`, `.cache` et al) it will never get a match.

With that out of the way here is a patch you can apply to fix it using https://yarnpkg.com/features/protocols/#patch
```diff
diff --git a/bin/cli.js b/bin/cli.js
index 6640b376d7f449e8551eb970cadf267784224adb..b34fc4117952c01b7d630fbc3053fa73c855be9f 100644
--- a/bin/cli.js
+++ b/bin/cli.js
@@ -17,7 +17,7 @@ const chalk = require('chalk');
 const isGitClean = require('is-git-clean');
 
 const transformerDirectory = path.join(__dirname, '../', 'transforms');
-const jscodeshiftExecutable = require.resolve('.bin/jscodeshift');
+const jscodeshiftExecutable = require.resolve('jscodeshift/bin/jscodeshift.js');
 
 function checkGitStatus(force) {
   let clean = false;
@@ -113,7 +113,7 @@ function runTransform({ files, flags, parser, transformer, answers }) {
 
   console.log(`Executing command: jscodeshift ${args.join(' ')}`);
 
-  const result = execa.sync(jscodeshiftExecutable, args, {
+  const result = execa.sync('node', [jscodeshiftExecutable, ...args], {
     stdio: 'inherit',
     stripEof: false
   });

```

**Reply by merceyz** [reply]
Made a PR upstream - https://github.com/reactjs/react-codemod/pull/295
Made one to https://github.com/skovhus/jest-codemods/pull/214 as well since that's where `react-codemod` got it from

**Reply by wdfinch** [reply]
Awesome thanks so much. The patch feature is really cool! I didn't know about that one.
